clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3490,6 +3490,138 @@ void handleFailure(Throwable throwable) {
+            tryComplete();
+        }
+
+        private synchronized void tryComplete() {
+            if (remaining.isEmpty()) {
+                ArrayList<Object> results = new ArrayList<>(listings.values());
+                results.addAll(errors);
+                future.complete(results);
+            }
+        }
+    }
+
+    @Override
+    public ListGroupsResult listGroups(ListGroupsOptions options) {
+        final KafkaFutureImpl<Collection<Object>> all = new KafkaFutureImpl<>();
+        final long nowMetadata = time.milliseconds();
+        final long deadline = calcDeadlineMs(nowMetadata, options.timeoutMs());
+        runnable.call(new Call("findAllBrokers", deadline, new LeastLoadedNodeProvider()) {
+            @Override
+            MetadataRequest.Builder createRequest(int timeoutMs) {
+                return new MetadataRequest.Builder(new MetadataRequestData()
+                    .setTopics(Collections.emptyList())
+                    .setAllowAutoTopicCreation(true));
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                MetadataResponse metadataResponse = (MetadataResponse) abstractResponse;
+                Collection<Node> nodes = metadataResponse.brokers();
+                if (nodes.isEmpty())
+                    throw new StaleMetadataException("Metadata fetch failed due to missing broker list");
+
+                HashSet<Node> allNodes = new HashSet<>(nodes);
+                final ListGroupsResults results = new ListGroupsResults(allNodes, all);
+
+                for (final Node node : allNodes) {
+                    final long nowList = time.milliseconds();
+                    runnable.call(new Call("listGroups", deadline, new ConstantNodeIdProvider(node.id())) {
+                        @Override
+                        ListGroupsRequest.Builder createRequest(int timeoutMs) {
+                            List<String> groupTypes = options.types()
+                                .stream()
+                                .map(GroupType::toString)
+                                .collect(Collectors.toList());
+                            return new ListGroupsRequest.Builder(new ListGroupsRequestData()
+                                .setTypesFilter(groupTypes)
+                            );
+                        }
+
+                        private void maybeAddGroup(ListGroupsResponseData.ListedGroup group) {
+                            final String groupId = group.groupId();
+                            final GroupType type = group.groupType().isEmpty()
+                                ? GroupType.UNKNOWN
+                                : GroupType.parse(group.groupType());
+                            final String protocolType = group.protocolType();
+                            final GroupListing groupListing = new GroupListing(
+                                groupId,
+                                Optional.of(type),
```

Comment:
It seems that `type` is never null, so why do we need to wrap it in `Optional`?

Reply:
The KIP reflects the fact that ListGroupResponse can omit the group type, so I was expecting this case to be `Optional.empty()`. However, in practice, I realise that `GroupType.UNKNOWN` does the job quite well.

I think there are two choices here. For a missing group type, use Optional.empty(), or use GroupType.UNKNOWN, in which case, the presence of `Optional` is not helpful and I should probably change the method signature to remove it. wdyt?

Reply:
I prefer using `Optional` instead of `GroupType.UNKNOWN`, since `GroupType` is part of `ListGroupsOptions`. The following usage seems a bit awkward to me.
```java
new ListGroupsOptions()
  .withTypes(Set.of(GroupType.UNKNOWN));
```

Reply:
This new code is part of AK 4.0. If you use this new admin client to list groups for a broker which does not support ListGroups v5 or later, it cannot return the group type. In this case, I think the `GroupListing` should use `Optional.empty()` for the group type.

I don't have a problem with `ListGroupsOptions.withTypes` not being able to request information on groups whose type cannot be returned. Using a set of `Optional` is cumbersome for a situation which is unlikely to be used in practice.

If you use this admin client to list groups for a future broker that has more group types than this admin client understands, I think the `GroupListing` should use `Optional.of(GroupType.UNKNOWN)` for the group type.

I think it's very unlikely that a user would want to query groups whose type cannot be understood. Technically, `GroupType.UNKNOWN` exists so it could be supplied in the `withTypes` method. It's not a sensible thing to attempt to do. I wonder whether the best thing would be to ensure that `GroupType.UNKNOWN` is not specified in the set of group types.

Reply:
@AndrewJSchofield thanks for this nice conclusion. Could you please add it to docs of `GroupListing#type`?